sycl/doc/extensions/proposed/sycl_ext_oneapi_forward_progress.asciidoc

```diff
@@ -0,0 +1,555 @@
+
+[NOTE]
+====
+Qualifying a thread of execution's forward progress guarantees with a coordination
+scope can be thought of as a shorthand for combining the forward progress
+guarantees provided by each thread of execution in the hierarchy.
+For example, consider a hypothetical implementation that creates a separate
+`std::thread` (providing concurrent forward progress guarantees) to execute
+each sub-group, but which permits an unbounded number of work-groups executed
+as tasks (providing parallel forward progress guarantees). With work-group
+coordination scope, each sub-group provides concurrent forward progress
+guarantees: this reflects the guarantees associated with each `std::thread`
+upon creation. With root-group coordination scope, each sub-group provides
+only parallel forward progress guarantees: each sub-group's progress guarantees
+are weakened to reflect that the creation of a sub-group depends upon a
+specific work-group (task) being scheduled and making progress.
+====
+
+
+=== Forward progress guarantees
+
+The `forward_progress_guarantee` `enum` is used to represent the three classes
+of forward progress guarantee defined in the {cpp} core language specification.
+
+[source,c++]
+----
+namespace ext::oneapi::experimental::sycl {
+
+enum class forward_progress_guarantee {
+  concurrent,
+  parallel,
+  weakly_parallel
+};
+
+}
+----
+
+[NOTE]
+====
+This `enum` is defined similarly to the one defined in
+https://wg21.link/p2300[P2300: `std::execution`].
+====
+
+
+=== Execution scopes
+
+The `execution_scope` `enum` is used to represent the hierarchy of threads of
+execution within a SYCL implementation.
+
+[source,c++]
+----
+namespace ext::oneapi::experimental::sycl {
+
+enum class execution_scope {
+  work_item,
+  sub_group,
+  work_group,
+  root_group,
+};
+
+}
+----
+
+[NOTE]
+====
+The existing `memory_scope` `enum` is not used here to provide greater
+flexibility in modifying the forward progress extension in future.
+`memory_scope` is lacking values required by this extension (i.e. `root_group`)
+and has additional values that are not required here (e.g. `system`).
+====
+
+
+=== Device queries
+
+New device queries are introduced to allow developers to reason about the
+progress guarantees that can be provided by each device.
+
+The results of these queries represent the strongest forward progress
+guarantees that a device can satisfy, and therefore the strongest forward
+progress guarantees that can be requested by a kernel's <<properties,
+properties>>.
+
+The queries are expressed in terms of a _target_ execution scope and a
+_coordination scope_, as described previously. The _coordination_ scope
+must be broader than the _target_ execution scope. Note that these queries
+cannot be used to reason about the thread of execution associated with
+the root-group, because it is always the top of the hierarchy; when the
+target execution scope is `root_group`, it is not possible to specify
+a broader coordination scope.
```

Comment:
Good point.  If we're going to allow people to modify the hierarchy, there's no reason to prevent them going above the root-group.  I've moved things to a non-normative note in 5f382df2b6b8a4fe65fe776168df65cdc1416d58, and weakened slightly to say that the root-group is _currently_ the top of the hierarchy.